datafusion/functions-window/src/ntile.rs

```diff
@@ -0,0 +1,201 @@
+use datafusion_physical_expr_common::physical_expr::PhysicalExpr;
+use field::WindowUDFFieldArgs;
+
+get_or_init_udwf!(
+    Ntile,
+    ntile,
+    "integer ranging from 1 to the argument value, dividing the partition as equally as possible",
+    Ntile::create
+);
+
+pub fn ntile(arg: i64) -> datafusion_expr::Expr {
+    ntile_udwf().call(vec![arg.lit()])
+}
+
+#[derive(Debug)]
+pub struct Ntile {
+    signature: Signature,
+}
+
+impl Ntile {
+    /// Create a new `ntile` function
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::uniform(
+                1,
+                vec![
+                    DataType::UInt64,
+                    DataType::UInt32,
+                    DataType::UInt16,
+                    DataType::UInt8,
+                    DataType::Int64,
+                    DataType::Int32,
+                    DataType::Int16,
+                    DataType::Int8,
+                ],
+                Volatility::Immutable,
+            ),
+        }
+    }
+
+    pub fn create() -> Self {
+        Self::new()
+    }
+}
+
+static DOCUMENTATION: OnceLock<Documentation> = OnceLock::new();
+
+fn get_ntile_doc() -> &'static Documentation {
+    DOCUMENTATION.get_or_init(|| {
+        Documentation::builder()
+            .with_doc_section(DOC_SECTION_RANKING)
+            .with_description(
+                "Integer ranging from 1 to the argument value, dividing the partition as equally as possible",
+            )
+            .with_syntax_example("ntile(expression)")
+            .with_argument("expression","An integer describing the number groups the partition should be split into")
+            .build()
+            .unwrap()
+    })
+}
+
+impl WindowUDFImpl for Ntile {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "ntile"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn expressions(&self, expr_args: ExpressionArgs) -> Vec<Arc<dyn PhysicalExpr>> {
+        parse_expr(expr_args.input_exprs(), expr_args.input_types())
+            .into_iter()
+            .collect::<Vec<_>>()
+    }
+
+    fn partition_evaluator(
+        &self,
+        partition_evaluator_args: PartitionEvaluatorArgs,
+    ) -> Result<Box<dyn PartitionEvaluator>> {
+        let scalar_n =
+            get_scalar_value_from_args(partition_evaluator_args.input_exprs(), 0)?
+                .ok_or_else(|| {
+                    DataFusionError::Execution(
+                        "NTILE requires a positive integer".to_string(),
+                    )
+                })?;
+
+        let n = get_unsigned_integer(scalar_n)?;
+
+        Ok(Box::new(NtileEvaluator { n }))
+    }
+    fn field(&self, field_args: WindowUDFFieldArgs) -> Result<Field> {
+        let nullable = false;
+        let return_type: &DataType = field_args.input_types().first().unwrap();
+
+        Ok(Field::new(self.name(), return_type.clone(), nullable))
```

Comment:
When constructing the result field (column) we need to pass the fully qualified name in the schema which is available in `WindowUDFFieldArgs`.
https://github.com/apache/datafusion/blob/818ce3f01efe1213a9a1eda5dff1542bb9d457f7/datafusion/expr/src/udwf.rs#L387-L390

It is easy to get confused with the name of the udwf 😅. 

> 
> ```
> External error: query failed: DataFusion error: Internal error: Input field name ntile does not match with the projection expression ntile(Int64(8)) ORDER BY [aggregate_test_100.c4 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW.
> This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
> [SQL] SELECT
> NTILE(8) OVER (ORDER BY C4) as ntile1,
> NTILE(12) OVER (ORDER BY C12 DESC) as ntile2
> FROM aggregate_test_100
> ORDER BY c7
> LIMIT 5
> at test_files/window.slt:743
> ```

```suggestion
        Ok(Field::new(field_args.name(), return_type.clone(), nullable))
```